codegen/src/api/errors.rs

```diff
@@ -0,0 +1,171 @@
+/// variants, so that we can convert `u8` error codes inside a generated `DispatchError` into
+/// nicer error strings with documentation. To do this, we emit the type we'll return instances of,
+/// and a function that returns such an instance for all of the error codes seen in the metadata.
+pub fn generate_error_details(metadata: &RuntimeMetadataV14) -> ErrorDetails {
+    let errors = match pallet_errors(metadata) {
+        Ok(errors) => errors,
+        Err(e) => {
+            let err_string =
+                format!("Failed to generate error details from metadata: {}", e);
+            return ErrorDetails::emit_compile_error(&err_string)
+        }
+    };
+
+    let match_body_items = errors.into_iter().map(|err| {
+        let docs = err.description();
+        let pallet_index = err.pallet_index;
+        let error_index = err.error_index;
+        let pallet_name = err.pallet;
+        let error_name = err.error;
+
+        quote! {
+            (#pallet_index, #error_index) => Some(ErrorDetails {
+                pallet: #pallet_name,
+                error: #error_name,
+                docs: #docs
+            })
+        }
+    });
+
+    ErrorDetails {
+        // A type we'll be returning that needs defining at the top level:
+        type_def: quote! {
+            pub struct ErrorDetails {
+                pub pallet: &'static str,
+                pub error: &'static str,
+                pub docs: &'static str,
+            }
+        },
+        // A function which will live in an impl block for our DispatchError,
+        // to statically return details for known error types:
+        dispatch_error_impl_fn: quote! {
+            pub fn details(&self) -> Option<ErrorDetails> {
+                if let Self::Module { index, error } = self {
+                    match (index, error) {
+                        #( #match_body_items ),*,
+                        _ => None
+                    }
+                } else {
+                    None
+                }
+            }
+        },
+    }
+}
+
+fn pallet_errors(
+    metadata: &RuntimeMetadataV14,
+) -> Result<Vec<ErrorMetadata>, InvalidMetadataError> {
+    let get_type_def_variant = |type_id: u32| {
+        let ty = metadata
+            .types
+            .resolve(type_id)
+            .ok_or(InvalidMetadataError::MissingType(type_id))?;
+        if let scale_info::TypeDef::Variant(var) = ty.type_def() {
+            Ok(var)
+        } else {
+            Err(InvalidMetadataError::TypeDefNotVariant(type_id))
+        }
+    };
+
+    let pallet_errors = metadata
+        .pallets
+        .iter()
+        .filter_map(|pallet| {
+            pallet.error.as_ref().map(|error| {
+                let type_def_variant = get_type_def_variant(error.ty.id())?;
+                Ok((pallet, type_def_variant))
+            })
+        })
+        .collect::<Result<Vec<(_, _)>, _>>()?;
+
+    let errors = pallet_errors
```

Comment:
Maybe we can unify these two chains into a single? Like, why collect the `pallet_errors` first and then build the `errors` Vec?

Reply:
It might be a bit tricky; We end up with `Result<_>` Items just prior to that collect, so the collecting I think is just there to give a chance to bail if we encounter any errors!

Reply:
I settled on standard for loops; it felt tidier:

```rust
let mut pallet_errors = vec![];
for pallet in &metadata.pallets {
    let error = match &pallet.error {
        Some(err) => err,
        None => continue
    };

    let type_def_variant = get_type_def_variant(error.ty.id())?;
    for var in type_def_variant.variants().iter() {
        pallet_errors.push(ErrorMetadata {
            pallet_index: pallet.index,
            error_index: var.index(),
            pallet: pallet.name.clone(),
            error: var.name().clone(),
            variant: var.clone(),
        });
    }
}
```